src/libasr/pass/intrinsic_function_registry.h

```diff
@@ -814,51 +1038,147 @@ static inline ASR::expr_t* instantiate_Any(Allocator &al, const Location &loc,
+        if (!ASRUtils::is_character(*expr_type(arg))) {
+            err("str.partition() takes one arguments of type: str", arg->base.loc);
+        }
+
+        Vec<ASR::expr_t *> e_args; e_args.reserve(al, 1);
+        e_args.push_back(al, s_var);
+        e_args.push_back(al, arg);
+
+        ASR::ttype_t *return_type = b.Tuple({character(-2), character(-2), character(-2)});
+        ASR::expr_t *value = nullptr;
+        if (ASR::is_a<ASR::StringConstant_t>(*s_var)
+         && ASR::is_a<ASR::StringConstant_t>(*arg)) {
+            std::string s_sep = ASR::down_cast<ASR::StringConstant_t>(arg)->m_s;
+            std::string s_str = ASR::down_cast<ASR::StringConstant_t>(s_var)->m_s;
+            if (s_sep.size() == 0) {
+                err("Separator cannot be an empty string", arg->base.loc);
+            }
+            value = eval_Partition(al, loc, s_str, s_sep);
+        }
+
+        return ASR::make_IntrinsicFunction_t(al, loc,
+            static_cast<int64_t>(ASRUtils::IntrinsicFunctions::Partition),
+            e_args.p, e_args.n, 0, return_type, value);
+    }
+
+    static inline ASR::expr_t *instantiate_Partition(Allocator &al,
+        const Location &loc, SymbolTable *scope,
+        Vec<ASR::ttype_t*>& /*arg_types*/, Vec<ASR::call_arg_t>& new_args,
+        int64_t /*overload_id*/, ASR::expr_t* compile_time_value)
+    {
+        // TODO: show runtime error for empty separator or pattern
+        declare_function_variables("_lpython_str_partition");
+        {
+            args.reserve(al, 2);
+            Variable(target_string, character(-2), In);
+            args.push_back(al, target_string);
+            Variable(pattern, character(-2), In);
+            args.push_back(al, pattern);
+        }
+        ASR::ttype_t *return_type = b.Tuple({character(-2), character(-2), character(-2)});
+        Variable(result, return_type, ReturnVar)
+        Variable(index, int32, Local);
+        body.push_back(al, Assignment(index, b.Call(UnaryIntrinsicFunction::
+            create_KMP_function(al, loc, scope), args, int32)));
+        body.push_back(al, b.If(iEq(index, i32_n(-1)), {
+                Assignment(result, b.TupleConstant({ args[0],
+                    StringConstant("", character(0)),
+                    StringConstant("", character(0)) },
+                b.Tuple({character(-2), character(0), character(0)})))
+            }, {
+                Assignment(result, b.TupleConstant({
+                    StringSection(args[0], i32(0), index), args[1],
+                    StringSection(args[0], iAdd(index, StringLen(args[1])),
+                        StringLen(args[0]))}, return_type))
+            }));
+        body.push_back(al, Return());
+        ASR::symbol_t *fn_sym = make_Function_t(fn_name, fn_symtab, dep, args,
+            body, result, Source, Implementation, nullptr);
+        scope->add_symbol(fn_name, fn_sym);
+        return b.Call(fn_sym, new_args, return_type, compile_time_value);
```

Comment:
@certik, during our last discussion, you suggested renaming `Assign` to `Assignment`.
Should I do the same for `Call` as well? `Call` -> `FunctionCall`?

Reply:
I would. This can be done later too.